### PR TITLE
Add vocabulary and embedding

### PR DESCRIPTION
**Pull request by astonzhang**
## Description ##

Add vocabulary and embedding

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] utils, (and when applicable, API doc)
- [x] vocab, (and when applicable, API doc)
- [x] embedding, (and when applicable, API doc)

## Comments ##
- NA

### CODE REVIEW

File: tests/python/unittest/test_text.py [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+    seq1 = token_delim.join(['a', '0.1', '0.2', '0.3', '0.4', '0.5']) + '\n'
+    seq2 = token_delim.join(['b', '0.6', '0.7', '0.8', '0.9', '1.0']) + '\n'
+    seq3 = token_delim.join(['c', '0.6', '0.7', '0.8']) + '\n'
+    seqs = seq1 + seq2 + seq3
+    with open(os.path.join(path, pretrain_file), 'w') as fout:
+        fout.write(seqs)
+
+
+def test_token_embedding_from_file():
+    embed_root = 'embedding'
```
Comment (marcoabreu) [inline tests/python/unittest/test_text.py]: Please use a tempfile instead

File: tests/python/unittest/test_text.py [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+    reg = text.embedding.get_file_names(embedding_name=None)
+
+    assert len(reg['glove']) == 10
+    assert len(reg['fasttext']) == 327
+
+    assertRaises(KeyError, text.embedding.get_file_names, 'unknown$$')
+
+
+def test_vocab_set_embedding_with_one_custom_embedding():
+    embed_root = 'embedding'
```
Comment (marcoabreu) [inline tests/python/unittest/test_text.py]: Please use a tempfile instead

File: tests/python/unittest/test_text.py [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+                                  [1, 1, 1, 1, 1],
+                                  [1, 1, 1, 1, 1],
+                                  [3, 3, 3, 3, 3],
+                                  [2, 2, 2, 2, 2],
+                                  [1, 1, 1, 1, 1]])
+                        )
+
+
+def test_vocabulary_with_two_custom_embeddings():
+    embed_root = '.'
```
Comment (marcoabreu) [inline tests/python/unittest/test_text.py]: Please use a tempfile instead

File: python/mxnet/text/embedding.py [hunk 4]
```diff
@@ -0,0 +1,583 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# coding: utf-8
+# pylint: disable=consider-iterating-dictionary
+# pylint: disable=super-init-not-called
+# pylint: disable=arguments-differ
```
Comment (marcoabreu) [inline python/mxnet/text/embedding.py]: Are the last two pylint ignores really invalid?

**Comment by piiswrong** [comment]
This should be in gluon.
If we put vocab and embedding in mxnet.text and textdataset in gluon it's going to be really confusing

File: python/mxnet/text/embedding.py [hunk 5]
```diff
@@ -38,8 +38,12 @@
...
+
+
     Once an embedding is registered, we can create an instance of this embedding with
-    :func:`~mxnet.contrib.text.embedding.create`.
+    :func:`~mxnet.text.embedding.create`.
+
+
     Examples
     --------
     >>> @mxnet.contrib.text.embedding.register
```
Comment (kobenaxie) [inline python/mxnet/text/embedding.py]: `>>> @mxnet.text.embedding.register`

Comment (astonzhang) [reply python/mxnet/text/embedding.py]: Thanks!

File: python/mxnet/gluon/text/embedding.py [hunk 6]
```diff
@@ -29,10 +29,10 @@
 import warnings
 import zipfile
 
-from . import _constants as C
-from mxnet import ndarray as nd
+from mxnet import nd
```
Comment (szha) [inline python/mxnet/gluon/text/embedding.py]: we usually use relative import.

File: docs/api/python/gluon/text.md [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+>>> fasttext = text.embedding.create('fasttext', file_name='wiki.simple.vec')
+
+```
+
+Now, suppose that we have a simple text data set in the string format. We can count word frequency
+in the data set.
+
+```python
+>>> text_data = " hello world \n hello nice world \n hi world \n"
+>>> counter = text.count_tokens_from_str(text_data)
```
Comment (szha) [inline docs/api/python/gluon/text.md]: It doesn't seem necessary to create vocab just to access embedding vector.

File: docs/api/python/gluon/text.md [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+
+```
+
+The obtained `counter` has key-value pairs whose keys are words and values are word frequencies.
```
Comment (szha) [inline docs/api/python/gluon/text.md]: Should explain why a counter is needed first.

File: python/mxnet/gluon/text/embedding.py [hunk 9]
```diff
@@ -0,0 +1,582 @@
...
+        """Creates a user-defined token embedding from a pre-trained embedding file.
+
+
+        This is to load embedding vectors from a user-defined pre-trained token embedding file.
+        Denote by '(ed)' the argument `elem_delim`. Denote by (v_ij) the j-th element of the token
+        embedding vector for (token_i), the expected format of a custom pre-trained token embedding
+        file is:
+
+        '(token_1)(ed))v_11)(ed)(v_12)(ed)...(ed)(v_1k)\\\\n
+        (token_2)(ed)(v_21)(ed)(v_22)(ed)...(ed)(v_2k)\\\\n...'
```
Comment (szha) [inline python/mxnet/gluon/text/embedding.py]: Use an example for the file format inside a code block so that it's easier to understand the file format. Currently it looks confusing. http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-10074/10/api/python/gluon/text.html#mxnet.gluon.text.embedding.TokenEmbedding.from_file

File: python/mxnet/gluon/text/vocab.py [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+        A dict mapping each token to its index integer.
+    unknown_token : hashable object
+        The representation for any unknown token. In other words, any unknown token will be indexed
+        as the same representation.
+
+
+    Examples
+    --------
+    >>> fasttext = text.embedding.create('fasttext', file_name='wiki.simple.vec')
+    >>> text_data = " hello world \n hello nice world \n hi world \n"
```
Comment (szha) [inline python/mxnet/gluon/text/vocab.py]: Formatting is broken. http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-10074/10/api/python/gluon/text.html#mxnet.gluon.text.vocab.Vocabulary

File: python/mxnet/gluon/text/vocab.py [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+            return self._token_to_idx[tokens] if tokens in self._token_to_idx \
+                   else C.UNKNOWN_IDX
+        else:
+            return [self._token_to_idx[token] if token in self._token_to_idx
+                    else C.UNKNOWN_IDX for token in tokens]
+
+    def __len__(self):
+        return len(self._idx_to_token)
+
+    def set_embedding(self, embeddings):
```
Comment (szha) [inline python/mxnet/gluon/text/vocab.py]: Use `(self, *embeddings)` instead. `embeddings` should not be a list. After the change, it should be possible to do:
```python
vocab.set_embedding(fasttext_emb, glove_embed)
```

Comment (szha) [reply python/mxnet/gluon/text/vocab.py]: Remember to update the doc/example accordingly.

File: docs/api/python/gluon/text.md [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+4
+>>> my_vocab[['hello', 'world']]
+[3, 2]
+```
+
+Besides the specified unknown token '(unk)' and reserved_token '(pad)' are indexed, the 2 most
+frequent words 'world' and 'hello' are also indexed.
+
+
+### Assign token embedding to vocabulary
```
Comment (szha) [inline docs/api/python/gluon/text.md]: Assign doesn't seem like the right verb. Maybe attach?

File: python/mxnet/gluon/text/vocab.py [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+    def embedding(self):
+        return self._embedding
+
+    @property
+    def idx_to_token(self):
+        return self._idx_to_token
+
+    @property
+    def reserved_tokens(self):
+        return self._reserved_tokens
```
Comment (szha) [inline python/mxnet/gluon/text/vocab.py]: Should the `reserved_tokens` property always include `unknown_token`, given that they are all indexed first?

File: docs/api/python/gluon/text.md (hunk 12 above)
Comment (astonzhang) [reply docs/api/python/gluon/text.md]: Resolved

File: python/mxnet/gluon/text/vocab.py (hunk 10 above)
Comment (astonzhang) [reply python/mxnet/gluon/text/vocab.py]: resolved

File: python/mxnet/gluon/text/vocab.py (hunk 11 above)
Comment (astonzhang) [reply python/mxnet/gluon/text/vocab.py]: resolved. with updated test cases

File: docs/api/python/gluon/text.md (hunk 7 above)
Comment (astonzhang) [reply docs/api/python/gluon/text.md]: resolved

File: docs/api/python/gluon/text.md (hunk 8 above)
Comment (astonzhang) [reply docs/api/python/gluon/text.md]: resolved

File: python/mxnet/gluon/text/vocab.py (hunk 10 above)
Comment (astonzhang) [reply python/mxnet/gluon/text/vocab.py]: resolved

File: python/mxnet/gluon/text/embedding.py (hunk 9 above)
Comment (astonzhang) [reply python/mxnet/gluon/text/embedding.py]: resolved

File: python/mxnet/gluon/text/vocab.py (hunk 13 above)
Comment (astonzhang) [reply python/mxnet/gluon/text/vocab.py]: resolved with more detailed api specification.
